README
NAME
    DBIx::TempDB - Create a temporary database

VERSION
    0.13

SYNOPSIS
      use Test::More;
      use DBIx::TempDB;
      use DBI;

      # provide credentials with environment variables
      plan skip_all => 'TEST_PG_DSN=postgresql://postgres@localhost' unless $ENV{TEST_PG_DSN};

      # create a temp database
      my $tmpdb = DBIx::TempDB->new($ENV{TEST_PG_DSN});

      # print complete url to db server with database name
      diag $tmpdb->url;

      # useful for reading in fixtures
      $tmpdb->execute("create table users (name text)");
      $tmpdb->execute_file("path/to/file.sql");

      # connect to the temp database
      my $db = DBI->connect($tmpdb->dsn);

      # run tests...

      done_testing;
      # database is cleaned up when test exit

DESCRIPTION
    DBIx::TempDB is a module which allows you to create a temporary
    database, which only lives as long as your process is alive. This can be
    very convenient when you want to run tests in parallel, without messing
    up the state between tests.

    This module currently support PostgreSQL, MySQL and SQLite by installing
    the optional DBD::Pg, DBD::mysql and/or DBD::SQLite modules.

    Please create an issue <https://github.com/jhthorsen/dbix-tempdb/issues>
    or pull request for more backend support.

    This module is currently EXPERIMENTAL. That means that if any major
    design flaws have been made, they will be fixed without warning.

CAVEAT
    Creating a database is easy, but making sure it gets clean up when your
    process exit is a totally different ball game. This means that
    DBIx::TempDB might fill up your server with random databases, unless you
    choose the right "drop strategy". Have a look at the "drop_from_child"
    parameter you can give to "new" and test the different values and select
    the one that works for you.

ENVIRONMENT VARIABLES
  DBIX_TEMP_DB_KEEP_DATABASE
    Setting this variable will disable the core feature in this module: A
    unique database will be created, but it will not get dropped/deleted.

  DBIX_TEMP_DB_URL
    This variable is set by "create_database" and contains the complete URL
    pointing to the temporary database.

    Note that calling "create_database" on different instances of
    DBIx::TempDB will overwrite "DBIX_TEMP_DB_URL".

METHODS
  create_database
      $self = $self->create_database;

    This method will create a temp database for the current process. Calling
    this method multiple times will simply do nothing. This method is
    normally automatically called by "new".

    The database name generate is defined by the "template" parameter passed
    to "new", but normalization will be done to make it work for the given
    database.

  dsn
      ($dsn, $user, $pass, $attrs) = $self->dsn;
      ($dsn, $user, $pass, $attrs) = DBIx::TempDB->dsn($url);

    Will parse "url" or $url, and return a list of arguments suitable for
    "connect" in DBI.

    Note that this method cannot be called as an object method before
    "create_database" is called. You can on the other hand call it as a
    class method, with a URI::db or URL string as input.

  execute
      $self = $self->execute(@sql);

    This method will execute a list of @sql statements in the temporary SQL
    server.

  execute_file
      $self = $self->execute_file("relative/to/executable.sql");
      $self = $self->execute_file("/absolute/path/stmt.sql");

    This method will read the contents of a file and execute the SQL
    statements in the temporary server.

    This method is a thin wrapper around "execute".

  new
      $self = DBIx::TempDB->new($url, %args);
      $self = DBIx::TempDB->new("mysql://127.0.0.1");
      $self = DBIx::TempDB->new("postgresql://postgres@db.example.com");
      $self = DBIx::TempDB->new("sqlite:");

    Creates a new object after checking the $url is valid. %args can be:

    *   auto_create

        "create_database" will be called automatically, unless "auto_create"
        is set to a false value.

    *   create_database_command

        Can be set to a custom create database command in the database. The
        default is "create database %d", where %d will be replaced by the
        generated database name.

        For even more control, you can set this to a code ref which will be
        called like this:

          $self->$cb($database_name);

        The default is subject to change.

    *   drop_database_command

        Can be set to a custom drop database command in the database. The
        default is "drop database %d", where %d will be replaced by the
        generated database name.

        For even more control, you can set this to a code ref which will be
        called like this:

          $self->$cb($database_name);

        The default is subject to change.

    *   drop_from_child

        Setting "drop_from_child" to a true value will create a child
        process which will remove the temporary database, when the main
        process ends. There are two possible values:

        "drop_from_child=1" (the default) will create a child process which
        monitor the DBIx::TempDB object with a pipe. This will then DROP the
        temp database if the object goes out of scope or if the process
        ends.

        "drop_from_child=2" will create a child process detached from the
        parent, which monitor the parent with "kill(0, $parent)".

        The double fork code is based on a paste contributed by Easy Connect
        AS <http://easyconnect.no>, Knut Arne Bjørndal.

    *   template

        Customize the generated database name. Default template is
        "tmp_%U_%X_%H%i". Possible variables to expand are:

          %i = The number of tries if tries are higher than 0. Example: "_3"
          %H = Hostname
          %P = Process ID ($$)
          %T = Process start time ($^T)
          %U = UID of current user
          %X = Basename of executable

        The default is subject to change!

  url
      $url = $self->url;

    Returns the input URL as URI::db compatible object. This URL will have
    the dbname part set to the database from "create_database", but not
    *until* after "create_database" is actually called.

    The URL returned can be passed directly to modules such as Mojo::Pg and
    Mojo::mysql.

COPYRIGHT AND LICENSE
    Copyright (C) 2015, Jan Henning Thorsen

    This program is free software, you can redistribute it and/or modify it
    under the terms of the Artistic License version 2.0.

AUTHOR
    Jan Henning Thorsen - "jhthorsen@cpan.org"



Changes
Revision history for perl distribution DBIx-TempDB

0.16 Not Released
 - Add DBIx::TempDB::Util
 - Can clean up temp databases manually
 - Remove support custom create_database_command
 - Remove support custom drop_database_command

0.15 2018-07-08T18:57:16+0800
 - Guarantee that database names are lower case #9
   Contributor: Christopher Eveland <chris@eyeson.ai>

0.14 2016-10-25T22:48:20+0200
 - Removed EXPERIMENTAL notice

0.13 2016-09-08T20:10:34+0200
 - Skip tests on Windows

0.12 2016-04-17T21:33:34+0200
 - Change drop_from_child=1 is now default #8

0.11 2015-11-11T13:16:04+0100
 - Fix execute() to only pass one statement to $dbh->do()

0.10 2015-11-04T19:06:29+0100
 - Fix parsing "SET foreign_key_checks=0" MySQL statement

0.09 2015-10-07T16:59:12+0200
 - Will run cleanup from child on INT, QUIT and TERM

0.08 2015-09-22T14:19:21+0200
 - Fix running multiple commands in MySQL using execute()
   The parser code is taken from Mojo::mysql, written by Svetoslav Naydenov
 - Add support for custom "create_database_command"
 - Add support for custom "drop_database_command"
 - Allow execute of multiple statements by SQLite driver #6

0.07 2015-09-20T09:40:22+0200
 - Use URI::db instead of Mojo::URL #4
   Contributor: Dan Book
 - Fix duplicate "CREATE DATABASE x" statements fail less
 - Add warning when DBIX_TEMP_DB_KEEP_DATABASE is set

0.06 2015-09-18T14:52:06+0200
 - Add support for DBIX_TEMP_DB_KEEP_DATABASE=1

0.05 2015-09-18T14:00:30+0200
 - Add support for SQLite

0.04 2015-09-18T11:50:41+0200
 - Add support for "template"
 - Add drop_from_child=2, which monitor parent with a double forked process
 - Fix max length for database name is 63 characters (Pg=63, MySQL=64)

0.03 2015-09-17T21:59:16+0200
 - Add support for "drop_from_child"

0.02 2015-09-17T20:02:12+0200
 - Fix depending on DBI
 - Add create_database() sets DBIX_TEMP_DB_URL
 - Add support for calling dsn() as a class method
 - Add execute()
 - Add execute_file()

0.01 2015-09-17T19:12:10+0200
 - Can make temp database for postgresql
 - Can make temp database for mysql


Makefile.PL
# Generated by git-ship. See 'git-ship --man' for help or https://github.com/jhthorsen/app-git-ship
use utf8;
use ExtUtils::MakeMaker;
my %WriteMakefileArgs = (
  NAME           => 'DBIx::TempDB',
  AUTHOR         => 'Jan Henning Thorsen <jhthorsen@cpan.org>',
  LICENSE        => 'artistic_2',
  ABSTRACT_FROM  => 'lib/DBIx/TempDB.pm',
  VERSION_FROM   => 'lib/DBIx/TempDB.pm',
  EXE_FILES      => [qw()],
  OBJECT         => '',
  BUILD_REQUIRES => {},
  TEST_REQUIRES  => {'Test::More' => '0.88'},
  PREREQ_PM      => {'DBI' => '1.60', 'URI::QueryParam' => '1.69', 'URI::db' => '0.15', 'perl' => '5.010001'},
  META_MERGE     => {
    'dynamic_config' => 0,
    'meta-spec'      => {version => 2},
    'resources'      => {
      bugtracker => {web => 'https://github.com/jhthorsen/dbix-tempdb/issues'},
      homepage   => 'https://github.com/jhthorsen/dbix-tempdb',
      repository => {
        type => 'git',
        url  => 'https://github.com/jhthorsen/dbix-tempdb.git',
        web  => 'https://github.com/jhthorsen/dbix-tempdb',
      },
    },
    'x_contributors' => ['Jan Henning Thorsen <jhthorsen@cpan.org>'],
  },
  test => {TESTS => (-e 'META.yml' ? 't/*.t' : 't/*.t xt/*.t')},
);

unless (eval { ExtUtils::MakeMaker->VERSION('6.63_03') }) {
  my $test_requires = delete $WriteMakefileArgs{TEST_REQUIRES};
  @{$WriteMakefileArgs{PREREQ_PM}}{keys %$test_requires} = values %$test_requires;
}

WriteMakefile(%WriteMakefileArgs);


lib/DBIx/TempDB/Util.pm
package DBIx::TempDB::Util;
use strict;
use warnings;
use Exporter 'import';

use Carp qw(confess croak);
use POSIX;
use Scalar::Util 'blessed';
use URI::db;
use URI::QueryParam;

use constant DEBUG               => $ENV{DBIX_TEMP_DB_DEBUG}               || 0;
use constant KILL_SLEEP_INTERVAL => $ENV{DBIX_TEMP_DB_KILL_SLEEP_INTERVAL} || 2;

our @EXPORT_OK = qw(dsn_for on_process_end parse_sql);

sub dsn_for {
  my ($url, $database_name) = @_;
  $url = URI::db->new($url) unless blessed $url;
  croak "Unknown engine for $url" unless $url->has_recognized_engine;

  my $engine = $url->canonical_engine;
  $database_name //= $url->dbname;
  return _dsn_for_mysql($url, $database_name)  if $engine eq 'mysql';
  return _dsn_for_pg($url, $database_name)     if $engine eq 'pg';
  return _dsn_for_sqlite($url, $database_name) if $engine eq 'sqlite';
  croak "Can't create DSN for engine $engine.";
}

sub on_process_end {
  my $code = pop;
  my $mode = shift // 'fork';

  return _on_process_end_fork($code)        if $mode eq 'fork';
  return _on_process_end_double_fork($code) if $mode eq 'double_fork';
  return DBIx::TempDB::Guard->new($code, $$);
}

sub parse_sql {
  my ($type, $sql) = @_;
  $type = $type->canonical_engine if blessed $type;
  return _parse_mysql($sql) if $type eq 'mysql';
  return $sql;
}

sub _dsn_for_mysql {
  my ($url, $database_name) = @_;
  my %opt = %{$url->query_form_hash};
  my ($dsn, @userinfo);

  $url = URI::db->new($url);
  $url->dbname($database_name);
  $url->query(undef);
  $dsn      = $url->dbi_dsn;
  @userinfo = ($url->user, $url->password);

  $opt{AutoCommit}          //= 1;
  $opt{AutoInactiveDestroy} //= 1;
  $opt{PrintError}          //= 0;
  $opt{RaiseError}          //= 1;
  $opt{mysql_enable_utf8}   //= 1;

  return $dsn, @userinfo[0, 1], \%opt;
}

sub _dsn_for_pg {
  my ($url, $database_name) = @_;
  my %opt = %{$url->query_form_hash};
  my ($dsn, @userinfo);

  $url = URI::db->new($url);
  $url->dbname($database_name);
  $url->query(undef);
  if (my $service = delete $opt{service}) { $url->query_param(service => $service) }
  $dsn      = $url->dbi_dsn;
  @userinfo = ($url->user, $url->password);

  $opt{AutoCommit}          //= 1;
  $opt{AutoInactiveDestroy} //= 1;
  $opt{PrintError}          //= 0;
  $opt{RaiseError}          //= 1;

  return $dsn, @userinfo[0, 1], \%opt;
}

sub _dsn_for_sqlite {
  my ($url, $database_name) = @_;
  my %opt = %{$url->query_form_hash};

  $url = URI::db->new($url);
  $url->dbname($database_name);
  $url->query(undef);
  my $dsn = $url->dbi_dsn;

  $opt{AutoCommit}          //= 1;
  $opt{AutoInactiveDestroy} //= 1;
  $opt{PrintError}          //= 0;
  $opt{RaiseError}          //= 1;
  $opt{sqlite_unicode}      //= 1;

  return $dsn, "", "", \%opt;
}

sub _on_process_end_double_fork {
  my $code = shift;
  my $ppid = $$;

  warn "[TempDB:$$] Watching process using double fork.\n" if DEBUG;
  local $SIG{CHLD} = 'DEFAULT';
  pipe(my $READER, my $WRITER) or confess "Couldn't create pipe: $!";

  # Parent
  if (my $pid_1 = fork // confess "Couldn't fork: $!") {
    my $pid_2;

    # Wait around until the second fork is done so that when we return from
    # here there are no new child processes that could mess things up if the
    # calling process does any process handling.
    close $WRITER;
    $pid_2 = <$READER>;
    $pid_2 = $pid_2 =~ m!(\d+)! ? $1 : undef;
    waitpid $pid_1, 0;
    confess "Couldn't get pid_2 from $pid_1." unless $pid_2;
    warn "[TempDB:$$] Double forked from $$ to $pid_1 to $pid_2\n" if DEBUG;
    return DBIx::TempDB::Guard->new(sub { kill TERM => $pid_2 }, $pid_2);
  }

  # Child #1
  # Detach completely from parent by creating our own session and process
  # group, closing all filehandles and forking a second time.
  $0 = "dbix-on-process-end-$ppid";
  close $READER;
  $DB::CreateTTY = 0;
  POSIX::setsid() != -1 or die "[TempDB:$$] Couldn't become session leader: $!\n";

  if (my $pid_2 = fork // die "[TempDB:$$] Couldn't fork: $!") {
    print $WRITER "$pid_2\n";
    close $WRITER;
    POSIX::_exit(0);
  }

  # Child #2
  warn "[TempDB:$ppid/$$] Double fork waiting on signals or parent to go away.\n" if DEBUG;
  _on_process_signals($code);
  sleep KILL_SLEEP_INTERVAL while kill 0, $ppid;
  local $ENV{DBIX_TEMP_DB_SIGNAL} = 'parent';
  $code->();
  exit;
}

sub _on_process_end_fork {
  my $code = shift;
  my $ppid = $$;

  # Parent
  warn "[TempDB:$$] Watching process using single fork.\n" if DEBUG;
  pipe(my $READER, my $WRITER) or confess "Couldn't create pipe: $!";
  defined(my $pid = fork)      or confess "Couldn't fork: $!";
  return DBIx::TempDB::Guard->new(sub { close $WRITER }, $pid) if $pid;

  # Child
  $0             = "dbix-on-process-end-$ppid";
  $DB::CreateTTY = 0;
  close $WRITER;
  warn "[TempDB:$ppid/$$] Fork waiting on signals or pipe to go away.\n" if DEBUG;
  _on_process_signals($code);
  1 while my $l = readline <$READER>;
  local $ENV{DBIX_TEMP_DB_SIGNAL} = 'pipe';
  $code->();
  exit;
}

sub _on_process_signals {
  my $code = shift;
  for my $name (qw(INT QUIT TERM)) {
    $SIG{$name} = sub { local $ENV{DBIX_TEMP_DB_SIGNAL} = $name; $code->(); exit; };
  }
}

sub _parse_mysql {
  my $sql = shift;
  my ($new, $last, $delimiter) = (0, '', ';');
  my @commands;

  while (length($sql) > 0) {
    my $token;

    if ($sql =~ /^$delimiter/x) {
      ($new, $token) = (1, $delimiter);
    }
    elsif ($sql =~ /^delimiter\s+(\S+)\s*(?:\n|\z)/ip) {
      ($new, $token, $delimiter) = (1, ${^MATCH}, $1);
    }
    elsif ($sql =~ /^(\s+)/s or $sql =~ /^(\w+)/) {    # general name
      $token = $1;
    }
    elsif (
      $sql    =~ /^--.*(?:\n|\z)/p                             # double-dash comment
      or $sql =~ /^\#.*(?:\n|\z)/p                             # hash comment
      or $sql =~ /^\/\*(?:[^\*]|\*[^\/])*(?:\*\/|\*\z|\z)/p    # C-style comment
      or $sql =~ /^'(?:[^'\\]*|\\(?:.|\n)|'')*(?:'|\z)/p       # single-quoted literal text
      or $sql =~ /^"(?:[^"\\]*|\\(?:.|\n)|"")*(?:"|\z)/p       # double-quoted literal text
      or $sql =~ /^`(?:[^`]*|``)*(?:`|\z)/p
      )
    {                                                          # schema-quoted literal text
      $token = ${^MATCH};
    }
    else {
      $token = substr($sql, 0, 1);
    }

    # chew token
    substr $sql, 0, length($token), '';

    if ($new) {
      push @commands, $last if $last !~ /^\s*$/s;
      ($new, $last) = (0, '');
    }
    else {
      $last .= $token;
    }
  }

  push @commands, $last if $last !~ /^\s*$/s;
  return map { s/^\s+//; $_ } @commands;
}

package DBIx::TempDB::Guard;
sub new     { my $class = shift; bless [@_], $class }
sub DESTROY { shift->[0]->() }

package DBIx::TempDB::Util;
1;

=encoding utf8

=head1 NAME

DBIx::TempDB::Util - Utility functions for DBIx::TempDB

=head1 SYNOPSIS

  use DBIx::TempDB::Util qw(dsn_for parse_sql);

  my $url = URI::db->new("postgresql://postgres@localhost");
  print join ", ", dsn_for($url);

  my $guard = on_process_end sub { ... };
  undef $guard; # call the code block earlier

  print $_ for parse_sql("mysql", "delimiter //\ncreate table y (bar varchar(255))//\n");

=head1 DESCRIPTION

L<DBIx::TempDB::Util> contains some utility functions for L<DBIx::TempDB>.

=head1 FUNCTIONS

=head2 dsn_for

  @dsn = dsn_for +URI::db->new("postgresql://postgres@localhost");
  @dsn = dsn_for "postgresql://postgres@localhost";

L</dsn_for> takes either a string or L<URI::db> object and returns a list of
arguments suitable for L<DBI/connect>.

=head2 on_process_end

  $guard = on_process_end sub { ... };
  $guard = on_process_end $mode       => sub { ... };
  $guard = on_process_end destroy     => sub { ... };
  $guard = on_process_end double_fork => sub { ... };
  $guard = on_process_end fork        => sub { ... };

Used to set up a code block to be called when the process ends. The default
C<$mode> is "fork". The C<$guard> value can be used to call the code block
before the process ends:

  undef $guard; # call sub { ... }

=over 2

=item * destroy

This mode will call the callback when the current process ends normally. This
means that if the process is killed with SIGKILL (9) or another untrapped
signal, then the callback will I<not> be called.

=item * double_fork

This mode will create a process that is detached from the parent process. The
double forked process will check if the parent is running by sending C<kill 0>
every two seconds. This mode might not be supported by all operating systems.

=item * fork

This mode will create a process with a pipe connected to the parent process.
Once the pipe is closed (when the parents ends) the callback will be called.
This should work in most processes, but will not work if a the process group
receives an unhandled signal.

=back

=head2 parse_sql

  @statements = parse_sql $type, $sql;
  @statements = parse_sql $uri_db, $sql;
  @statements = parse_sql "mysql", "insert into ...";

Takes either a string or an L<URI::db> object and a string containing SQL and
splits the SQL into a list of individual statements.

Currently only "mysql" is a supported type, meaning any other type will simply
return the input C<$sql>.

This is not required for SQLite though, you can do this instead:

  local $dbh->{sqlite_allow_multiple_statements} = 1;
  $dbh->do($sql);

=head1 SEE ALSO

L<DBIx::TempDB>.

=cut


lib/DBIx/TempDB.pm
package DBIx::TempDB;
use strict;
use warnings;

use Carp qw(confess croak);
use Cwd ();
use DBI;
use DBIx::TempDB::Util qw(dsn_for on_process_end parse_sql);
use File::Basename ();
use File::Spec;
use IO::Handle ();
use Scalar::Util 'blessed';
use Sys::Hostname ();
use URI::db;

use constant CWD                 => eval { File::Basename::dirname(Cwd::abs_path($0)) };
use constant DEBUG               => $ENV{DBIX_TEMP_DB_DEBUG} || 0;
use constant MAX_NUMBER_OF_TRIES => $ENV{DBIX_TEMP_DB_MAX_NUMBER_OF_TRIES} || 20;

our $VERSION         = '0.16';
our %SCHEMA_DATABASE = (pg => 'postgres', mysql => 'mysql', sqlite => '');
my $N = 0;

sub create_database {
  my $self = shift;
  return $self if $self->{created};

  local $@;
  my $mode = !$self->{drop_from_child} ? 'destroy' : $self->{drop_from_child} == 2 ? 'double_fork' : 'fork';
  my ($guard, $name) = 0;
  while (++$guard < MAX_NUMBER_OF_TRIES) {
    $name = $self->_generate_database_name($N + $guard - 1);
    eval { $self->_create_database($name) } or next;
    $self->{database_name} = $name;
    warn "[TempDB:$$] Created @{[$ENV{DBIX_TEMP_DB_KEEP_DATABASE} ? 'permanent' : 'temp']} database $name\n" if DEBUG;
    $self->{guard} = on_process_end $mode => $self->_drop_database_cb($self->{database_name})
      unless $ENV{DBIX_TEMP_DB_KEEP_DATABASE};
    $self->{created}++;
    $self->{url}->dbname($name);
    $ENV{DBIX_TEMP_DB_URL} = $self->{url}->uri->as_string;
    $N++;
    return $self;
  }

  croak qq(Couldn't create database "$name": $@);
}

sub drop_databases {
  my ($self, $params) = @_;

  my $self_db_name = $self->{database_name} || '';
  my $delete_self  = $params->{self}        || '';
  delete $self->{guard} if $delete_self;

  # Drop a single database by name
  return $self->_drop_database($params->{name}) if $params->{name};
  return $self->_drop_database($self_db_name)   if $delete_self eq 'only';

  # Drop sibling (and curren) databases
  my $max = $N > MAX_NUMBER_OF_TRIES ? $N : MAX_NUMBER_OF_TRIES;
  my @err;
  for my $n (0 .. $max) {
    my $name = $self->_generate_database_name($n);
    next unless $delete_self eq 'include' or ($self_db_name and $self_db_name ne $name);
    push @err, $@ unless eval { $self->_drop_database($name); 1 };
  }

  croak $err[0] if @err == $max;
  return $self;
}

sub dsn {
  my ($self, $url) = @_;

  unless (blessed $self) {
    Carp::carp("DBIx::TempDB->dsn(...) is deprecated. Use DBIx::TempDB::Util::dsn_for() instead");
    $url = URI::db->new($url) unless blessed $url;
    return dsn_for($url, $url->dbname);
  }

  croak "Can't call dsn() before create_database()" unless $self->{database_name};
  croak 'Database does not exist.' if $self->url->canonical_engine eq 'sqlite' and !-e $self->{database_name};
  return dsn_for($self->{url}, $self->{database_name});
}

sub execute {
  my $self = shift;
  my $dbh  = DBI->connect($self->dsn);
  local $dbh->{sqlite_allow_multiple_statements} = 1 if $self->url->canonical_engine eq 'sqlite';
  $dbh->do($_) for map { parse_sql($self->url, $_) } @_;
  return $self;
}

sub execute_file {
  my ($self, $path) = @_;

  unless (File::Spec->file_name_is_absolute($path)) {
    croak qq(Can't resolve path to "$path".) unless CWD;
    $path = File::Spec->catfile(CWD, split '/', $path);
  }

  open my $SQL, '<', $path or croak "Can't open $path: $!";
  my $ret = my $sql = '';
  while ($ret = $SQL->sysread(my $buffer, 131072, 0)) { $sql .= $buffer }
  croak qq{Can't read "$path": $!} unless defined $ret;
  warn "[TempDB:$$] Execute $path\n" if DEBUG;
  return $self->execute($sql);
}

sub new {
  my $class = shift;
  my $url   = URI::db->new(shift || '');
  my $self  = bless {@_, url => $url}, $class;

  $self->{drop_from_child} //= 1;
  $self->{schema_database} ||= $SCHEMA_DATABASE{$url->canonical_engine} // croak qq(Unsupported engine for $url);
  $self->{template}        ||= 'tmp_%U_%X_%H%i';
  warn "[TempDB:$$] schema_database=$self->{schema_database}\n" if DEBUG;

  return $self->create_database if $self->{auto_create} // 1;
  return $self;
}

sub url { shift->{url}->uri }

sub _create_database {
  my ($self, $name) = @_;

  if ($self->url->canonical_engine eq 'sqlite') {
    require IO::File;
    use Fcntl qw(O_CREAT O_EXCL O_RDWR);
    IO::File->new->open($name, O_CREAT | O_EXCL | O_RDWR) or confess "Can't write $name: $!\n";
  }
  else {
    DBI->connect($self->_schema_dsn)->do(sprintf 'create database %s', $name);
  }
}

sub _drop_database { shift->_drop_database_cb(shift)->() }

sub _drop_database_cb {
  my ($self, $name) = @_;

  if ($self->url->canonical_engine eq 'sqlite') {
    return sub {
      local $! = 0;
      unlink $name                                     if -e $name;
      confess "[TempDB:$$] Can't unlink $name: $!"     if $! and $! != 2;
      warn "[TempDB:$$] Dropped temp database $name\n" if DEBUG;
    };
  }

  my $sql = sprintf 'drop database if exists %s', $name;
  $sql =~ s!\%d!$name!g;
  return sub {
    my $dbh = DBI->connect($self->_schema_dsn);
    eval { $dbh->do('set client_min_messages to warning') };    # for postgres
    $dbh->do($sql);
    warn "[TempDB:$$] Dropped temp database $name\n" if DEBUG;
  };
}

sub _generate_database_name {
  my ($self, $n) = @_;
  my $name = $self->{template};

  $name =~ s/\%([iHPTUX])/{
      $1 eq 'i' ? ($n > 0 ? "_$n" : '')
    : $1 eq 'H' ? Sys::Hostname::hostname()
    : $1 eq 'P' ? $$
    : $1 eq 'T' ? $^T
    : $1 eq 'U' ? $<
    : $1 eq 'X' ? File::Basename::basename($0)
    :             "\%$1"
  }/egx;

  if (63 < length $name and !$self->{keep_too_long_database_name}) {
    confess qq(Can't create a shorter database name with "$self->{template}".)
      unless $self->{template} =~ s!\%T!!g
      or $self->{template}     =~ s!\%H!!g
      or $self->{template}     =~ s!\%X!!g;
    return $self->_generate_database_name($n);
  }

  $name =~ s!^/+!!;
  $name =~ s!\W!_!g;
  $name = lc $name;

  return $name if $self->url->canonical_engine ne 'sqlite';
  return File::Spec->catfile($self->_tempdir, "$name.sqlite");
}

sub _schema_dsn {
  my $self = shift;
  local $self->{database_name} = $self->{schema_database};
  return $self->dsn;
}

sub _tempdir {
  shift->{tempdir} ||= File::Spec->tmpdir;
}

1;

=encoding utf8

=head1 NAME

DBIx::TempDB - Create a temporary database

=head1 VERSION

0.16

=head1 SYNOPSIS

  use Test::More;
  use DBIx::TempDB;
  use DBI;

  # provide credentials with environment variables
  plan skip_all => 'TEST_PG_DSN=postgresql://postgres@localhost' unless $ENV{TEST_PG_DSN};

  # create a temp database
  my $tmpdb = DBIx::TempDB->new($ENV{TEST_PG_DSN});

  # print complete url to db server with database name
  diag $tmpdb->url;

  # useful for reading in fixtures
  $tmpdb->execute("create table users (name text)");
  $tmpdb->execute_file("path/to/file.sql");

  # connect to the temp database
  my $db = DBI->connect($tmpdb->dsn);

  # run tests...

  done_testing;
  # database is cleaned up when test exit

=head1 DESCRIPTION

L<DBIx::TempDB> is a module which allows you to create a temporary database,
which only lives as long as your process is alive. This can be very
convenient when you want to run tests in parallel, without messing up the
state between tests.

This module currently support PostgreSQL, MySQL and SQLite by installing the optional
L<DBD::Pg>, L<DBD::mysql> and/or L<DBD::SQLite> modules.

Please create an L<issue|https://github.com/jhthorsen/dbix-tempdb/issues>
or pull request for more backend support.

=head1 CAVEAT

Creating a database is easy, but making sure it gets clean up when your
process exit is a totally different ball game. This means that
L<DBIx::TempDB> might fill up your server with random databases, unless
you choose the right "drop strategy". Have a look at the L</drop_from_child>
parameter you can give to L</new> and test the different values and select
the one that works for you.

=head1 ENVIRONMENT VARIABLES

=head2 DBIX_TEMP_DB_KEEP_DATABASE

Setting this variable will disable the core feature in this module:
A unique database will be created, but it will not get dropped/deleted.

=head2 DBIX_TEMP_DB_URL

This variable is set by L</create_database> and contains the complete
URL pointing to the temporary database.

Note that calling L</create_database> on different instances of
L<DBIx::TempDB> will overwrite C<DBIX_TEMP_DB_URL>.

=head1 METHODS

=head2 create_database

  $tmpdb = $tmpdb->create_database;

This method will create a temp database for the current process. Calling this
method multiple times will simply do nothing. This method is normally
automatically called by L</new>.

The database name generate is defined by the L</template> parameter passed to
L</new>, but normalization will be done to make it work for the given database.

=head2 drop_databases

  $tmpdb->drop_databases;
  $tmpdb->drop_databases({self => "include"});
  $tmpdb->drop_databases({self => "only"});
  $tmpdb->drop_databases({name => "some_database_name"});

Used to drop either sibling databases (default), sibling databases and the
current database or a given database by name.

=head2 dsn

  ($dsn, $user, $pass, $attrs) = $tmpdb->dsn;

Will parse L</url> and return a list of arguments suitable for L<DBI/connect>.

Note that this method cannot be called as an object method before
L</create_database> is called.

See also L<DBIx::TempDB::Util/dsn_for>.

=head2 execute

  $tmpdb = $tmpdb->execute(@sql);

This method will execute a list of C<@sql> statements in the temporary
SQL server.

=head2 execute_file

  $tmpdb = $tmpdb->execute_file("relative/to/executable.sql");
  $tmpdb = $tmpdb->execute_file("/absolute/path/stmt.sql");

This method will read the contents of a file and execute the SQL statements
in the temporary server.

This method is a thin wrapper around L</execute>.

=head2 new

  $tmpdb = DBIx::TempDB->new($url, %args);
  $tmpdb = DBIx::TempDB->new("mysql://127.0.0.1");
  $tmpdb = DBIx::TempDB->new("postgresql://postgres@db.example.com");
  $tmpdb = DBIx::TempDB->new("sqlite:");

Creates a new object after checking the C<$url> is valid. C<%args> can be:

=over 4

=item * auto_create

L</create_database> will be called automatically, unless C<auto_create> is
set to a false value.

=item * drop_from_child

Setting "drop_from_child" to a true value will create a child process which
will remove the temporary database, when the main process ends. There are two
possible values:

C<drop_from_child=1> (the default) will create a child process which monitor
the L<DBIx::TempDB> object with a pipe. This will then DROP the temp database
if the object goes out of scope or if the process ends.

C<drop_from_child=2> will create a child process detached from the parent,
which monitor the parent with C<kill(0, $parent)>.

The double fork code is based on a paste contributed by
L<Easy Connect AS|http://easyconnect.no>, Knut Arne Bjørndal.

See also L<DBIx::TempDB::Util/on_process_end>.

=item * template

Customize the generated database name. Default template is "tmp_%U_%X_%H%i".
Possible variables to expand are:

  %i = The number of tries if tries are higher than 0. Example: "_3"
  %H = Hostname
  %P = Process ID ($$)
  %T = Process start time ($^T)
  %U = UID of current user
  %X = Basename of executable

The default is subject to change!

=back

=head2 url

  $url = $tmpdb->url;

Returns the input URL as L<URI::db> compatible object. This URL will have
the L<dbname|URI::db/dbname> part set to the database from L</create_database>,
but not I<until> after L</create_database> is actually called.

The URL returned can be passed directly to modules such as L<Mojo::Pg>
and L<Mojo::mysql>.

=head1 COPYRIGHT AND LICENSE

Copyright (C) 2015-2020, Jan Henning Thorsen

This program is free software, you can redistribute it and/or modify it under
the terms of the Artistic License version 2.0.

=head1 AUTHOR

Jan Henning Thorsen - C<jhthorsen@cpan.org>

=cut
